### PR TITLE
[full-ci][tests-only]cache yarn dependencies for web UI tests

### PR DESCRIPTION
**Pull request by SwikritiT**
## Description
This PR caches the yarn for webUITets so that it will be reused across the webUI pipelines

## Related Issue

- Fixes https://github.com/owncloud/ocis/issues/4047

## How Has This Been Tested?

- CI

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by pascalwengerter** [comment]
@SwikritiT I can recommend installing https://github.com/bazelbuild/buildtools/blob/master/buildifier/README.md and either manually run the linter or (preferably) enable lint-on-save :) otherwise those `.star` files can be quite a pain to handle!

**Comment by pascalwengerter** [comment]
> > @SwikritiT I can recommend installing https://github.com/bazelbuild/buildtools/blob/master/buildifier/README.md and either manually run the linter or (preferably) enable lint-on-save :) otherwise those `.star` files can be quite a pain to handle!
> 
> Thanks! Will do.

Maybe you can even bring this back to your team or add it to some setup guide, it's helpful across all oC projects (from my expierence)!

File: .drone.star [hunk 1]
```diff
@@ -249,6 +250,38 @@ def main(ctx):
...
+                 installWebTestRunner() +
+                 rebuildBuildArtifactCache(ctx, "tests-yarn", "/srv/app/web/tests/acceptance/.yarn"),
+        "trigger": {
+            "ref": [
+                "refs/heads/master",
+                "refs/tags/**",
+                "refs/pull/**",
+            ],
+        },
+        "volumes": [stepVolumeOC10Tests],
```
Comment (saw-jan) [inline .drone.star:268]:
```suggestion
        "volumes": [pipelineVolumeOC10Tests],
```

File: .drone.star [hunk 2]
```diff
@@ -249,6 +250,38 @@ def main(ctx):
...
+        "image": OC_CI_NODEJS % DEFAULT_NODEJS_VERSION,
+        "commands": [
+            ". /drone/src/.drone.env",
+            "git clone -b $WEB_BRANCH --single-branch --no-tags https://github.com/owncloud/web.git /srv/app/web",
+            "cd /srv/app/web",
+            "git checkout $WEB_COMMITID",
+            "cd tests/acceptance/",
+            "yarn install --immutable",
+        ],
+    }]
```
Comment (saw-jan) [inline .drone.star:283]:
```suggestion
        ],
        "volumes": [stepVolumeOC10Tests],
    }]
```

File: .drone.star [hunk 3]
```diff
@@ -249,6 +250,38 @@ def main(ctx):
...
     return pipelines
 
+def yarnCache(ctx):
+    return [{
+        "kind": "pipeline",
+        "type": "docker",
+        "name": "cache-yarn",
+        "steps": skipIfUnchanged(ctx, "cache") +
+                 installWebTestRunner() +
+                 rebuildBuildArtifactCache(ctx, "tests-yarn", "/srv/app/web/tests/acceptance/.yarn"),
```
Comment (saw-jan) [inline .drone.star:260]: This also needs to have a way to access shared volume `stepVolumeOC10Tests`.
I don't know if this step will get access to the shared volume. If not, you might have to change the function or clone web to the src directory directly and build cache from there.

**Review by saw-jan — COMMENTED** [review]
try the suggestion. If it doesn't work then you might need to either refactor `` function to take dependencies or clone web directly to the src directory (current dir)

**Comment by saw-jan** [comment]
> @SwikritiT I can recommend installing https://github.com/bazelbuild/buildtools/blob/master/buildifier/README.md and either manually run the linter or (preferably) enable lint-on-save :) otherwise those `.star` files can be quite a pain to handle!

Cool! :heart: 

I personally use vscode for everything and it also has a bezel extension and we can set up buildifier in it and auto-format starlark on save.
https://github.com/saw-jan/DocsCol/blob/main/format-starlark.md

File: .drone.star (hunk 1 above)
Comment (saw-jan) [inline .drone.star:268]:
```suggestion
        "volumes": [pipelineVolumeOC10Tests],
```

File: .drone.star (hunk 2 above)
Comment (saw-jan) [inline .drone.star:283]:
```suggestion
        ],
        "volumes": [stepVolumeOC10Tests],
    }]
```

File: .drone.star (hunk 3 above)
Comment (SwikritiT) [reply .drone.star:260]: other suggestions didn't work. This seems like the way to go

File: .drone.star [hunk 4]
```diff
@@ -704,39 +746,36 @@ def uiTestPipeline(ctx, filterTags, early_fail, runPart = 1, numberOfParts = 1,
...
+                         "commands": [
+                             "git clone -b master --depth=1 https://github.com/owncloud/testing.git /srv/app/testing",
+                             "cd /webTestRunner/tests/acceptance",
+                             "./run.sh",
+                         ],
+                     },
+                 ] + failEarly(ctx, early_fail),
         "services": selenium() + middlewareService(),
-        "volumes": [pipelineVolumeOC10Tests] +
+        "volumes": [pipelineVolumeOC10Tests, stepVolumeOC10Tests] +
```
Comment (saw-jan) [inline .drone.star:778]: pipeline and step volumes have different usage. Also, this change might not be needed.
```suggestion
        "volumes": [pipelineVolumeOC10Tests] +
```

File: .drone.star [hunk 5]
```diff
@@ -249,6 +250,47 @@ def main(ctx):
...
+        },
+    }]
+
+def installWebTestRunner():
+    return [{
+        "name": "install-web-test-runner",
+        "image": OC_CI_NODEJS % DEFAULT_NODEJS_VERSION,
+        "commands": [
+            ". /drone/src/.drone.env",
+            "git clone -b $WEB_BRANCH --single-branch --no-tags https://github.com/owncloud/web.git /webTestRunner",
```
Comment (saw-jan) [inline .drone.star:277]:
```suggestion
            "git clone -b $WEB_BRANCH --single-branch --no-tags https://github.com/owncloud/web.git webTestRunner",
```

File: .drone.star [hunk 6]
```diff
@@ -249,6 +250,47 @@ def main(ctx):
...
+        ],
+    }]
+
+def yarnInstallUITests():
+    return [{
+        "name": "yarn-install",
+        "image": OC_CI_NODEJS % DEFAULT_NODEJS_VERSION,
+        "commands": [
+            ". /drone/src/.drone.env",
+            "cd /webTestRunner",
```
Comment (saw-jan) [inline .drone.star:287]:
```suggestion
            "cd webTestRunner",
```

File: .drone.star [hunk 7]
```diff
@@ -704,39 +746,36 @@ def uiTestPipeline(ctx, filterTags, early_fail, runPart = 1, numberOfParts = 1,
...
+                             "LOCAL_UPLOAD_DIR": "/uploads",
+                             "NODE_TLS_REJECT_UNAUTHORIZED": 0,
+                             "RUN_PART": runPart,
+                             "DIVIDE_INTO_NUM_PARTS": numberOfParts,
+                             "EXPECTED_FAILURES_FILE": "/drone/src/tests/acceptance/expected-failures-webUI-on-%s-storage%s.md" % (storage.upper(), expectedFailuresFileFilterTags),
+                             "MIDDLEWARE_HOST": "http://middleware:3000",
+                         },
+                         "commands": [
+                             "git clone -b master --depth=1 https://github.com/owncloud/testing.git /srv/app/testing",
+                             "cd /webTestRunner/tests/acceptance",
```
Comment (saw-jan) [inline .drone.star:772]:
```suggestion
                             "cd webTestRunner/tests/acceptance",
```

**Review by saw-jan — COMMENTED** [review]
not in the root path but in the current workign dir (/drone/src)

File: .drone.star [hunk 8]
```diff
@@ -249,6 +250,48 @@ def main(ctx):
...
+def yarnInstallUITests():
+    return [{
+        "name": "yarn-install",
+        "image": OC_CI_NODEJS % DEFAULT_NODEJS_VERSION,
+        "commands": [
+            ". /drone/src/.drone.env",
+            "cd webTestRunner",
+            "git checkout $WEB_COMMITID",
+            "ls",
+            "cd webTestRunner/tests/acceptance/",
```
Comment (saw-jan) [inline .drone.star:290]: you are already in `webTestRunner` dir
```suggestion
            "cd tests/acceptance/",
```

File: .drone.star (hunk 4 above)
Comment (SwikritiT) [reply .drone.star]: that got left out. Removed now! Thanks for pointing it out!

**Review by saw-jan — COMMENTED** [review]
LGTM

**Review by phil-davis — APPROVED** [review]
LGTM

**Pull request merged**